src/frontend/src/optimizer/rule/over_agg_to_topn.rs

```diff
@@ -80,49 +80,77 @@ impl Rule for OverAggToTopNRule {
             predicate.clone().split_disjoint(&rank_col)
         };
 
-        // TODO: support multiple complex rank predicates.
         // Currently rank [ < | <= | > | >= ] N is used to implement group topn.
         // While rank = 1 is used to implement deduplication.
         let (limit, offset) = {
-            if rank_pred.conjunctions.len() != 1 {
-                tracing::error!("Multiple complex rank predicates is not supported yet.");
+            // rank >= lb
+            let mut lb = vec![];
+            // rank <= ub
+            let mut ub = vec![];
+            // rank == eq
+            let mut eq = vec![];
+
+            for cond in rank_pred.conjunctions {
+                if let Some((input_ref, cmp, v)) = cond.as_comparison_const() {
+                    assert_eq!(input_ref.index, window_func_pos);
+                    let v = v
+                        .cast_implicit(DataType::Int64)
+                        .ok()?
+                        .eval_row_const()
+                        .ok()??;
+                    let v = *v.as_int64();
+                    // Note: rank functions start from 1
+                    match cmp {
+                        ExprType::LessThanOrEqual => ub.push(v),
+                        ExprType::LessThan => ub.push(v - 1),
+                        ExprType::GreaterThan => lb.push(v + 1),
+                        ExprType::GreaterThanOrEqual => lb.push(v),
+                        _ => unreachable!(),
+                    }
+                } else if let Some((input_ref, v)) = cond.as_eq_const() {
+                    assert_eq!(input_ref.index, window_func_pos);
+                    let v = v
+                        .cast_implicit(DataType::Int64)
+                        .ok()?
+                        .eval_row_const()
+                        .ok()??;
+                    let v = *v.as_int64();
+                    eq.push(v);
+                } else {
+                    tracing::error!("Failed to optimize complex rank predicate {:?}", cond);
+                    return None;
+                }
+            }
+
+            if eq.len() > 1
+                || lb.len() > 1
+                || ub.len() > 1
+                || (!eq.is_empty() && (!lb.is_empty() || !ub.is_empty()))
+            {
+                tracing::error!("Failed to optimize multiple complex rank predicates");
```

Comment:
Just curious: why not just merging them together? For example, if `ub = [10, 20]`, then use `ub = 10`

Reply:
Just KISS. But yes it's not hard, so fixed now.